### PR TITLE
ci: add release job

### LINKED ISSUE #69: upload static shared library
Currently, libwasmvm_static.so is excluded from the remote repository by gitignore. Therefore, ci that uses static lib must be built locally each time. Let's upload the static lib to remote as well.

### PR DESCRIPTION
**Pull request by shiki-tak**
# Description

Closes #69 

Release test result: https://github.com/shiki-tak/wasmvm/releases/tag/v1.0.0-0.10.1

## Types of changes

- [ ] Bug fix (changes which fixes an issue)
- [ ] New feature (changes which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] ETC (build, ci, docs, perf, refactor, style, test)

## Checklist

- [ ] I followed the [contributing guidelines](https://github.com/line/wasmvm/blob/main/CONTRIBUTING.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] The PR title and commits are followed [conventional commit form](https://www.conventionalcommits.org/en/v1.0.0).

### CODE REVIEW

File: .github/workflows/builds_and_tests.yml [hunk 1]
```diff
@@ -134,6 +134,22 @@ jobs:
           name: libwasmvm
           path: ./api/libwasmvm.x86_64.so
 
+  build_static_lib:
+    needs: build_shared_library
```
Comment (tnasu) [inline .github/workflows/builds_and_tests.yml:138]: I'm not sure, but, is it needed?

File: .github/workflows/builds_and_tests.yml [hunk 2]
```diff
@@ -134,6 +134,22 @@ jobs:
...
+        uses: actions/setup-go@v2
+        with:
+          go-version: ^1.17.4
+        id: go
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Build docker image
+        run: cd builders && make docker-image-alpine
+      - name: Build static library for Alpine
+        run: make release-build-alpine
```
Comment (tnasu) [inline .github/workflows/builds_and_tests.yml:151]: Oh, this includes tests... How about leaving a comment about testing or updating the job name?

https://github.com/line/wasmvm/blob/main/Makefile#L74-L77

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -46,3 +79,20 @@ jobs:
           curl -s -H "Authorization: token ${GITHUB_TOKEN}" \
           -d "{\"ref\": \"refs/tags/${{ needs.get-version.outputs.package-version }}\", \"sha\": \"${GITHUB_SHA}\"}" \
           "https://api.github.com/repos/${GITHUB_REPOSITORY}/git/refs"
+
+  update-releases:
+    name: Update Latest release
+    needs:
+      - push-tag
```
Comment (tnasu) [inline .github/workflows/release.yml:86]: Is it enough `push-tag` only? I think `build_static_lib` is also needed.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -34,6 +34,39 @@ jobs:
...
+      - name: set up
+        uses: actions/setup-go@v2
+        with:
+          go-version: ^1.17.4
+        id: go
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Build docker image
+        run: cd builders && make docker-image-alpine
+      - name: Build static library for Alpine
```
Comment (tnasu) [inline .github/workflows/release.yml:54]: ditto

**Review by tnasu — CHANGES_REQUESTED** [review]

File: .github/workflows/builds_and_tests.yml (hunk 1 above)
Comment (shiki-tak) [reply .github/workflows/builds_and_tests.yml:138]: Yes, I want to test that each PR builds correctly.

File: .github/workflows/builds_and_tests.yml [hunk 5]
```diff
@@ -134,6 +134,22 @@ jobs:
...
+      - name: set up
+        uses: actions/setup-go@v2
+        with:
+          go-version: ^1.17.4
+        id: go
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Build docker image
+        run: cd builders && make docker-image-alpine
+      - name: Build static library for Alpine
```
[the remark this answers is not in the document]
Comment (shiki-tak) [reply .github/workflows/builds_and_tests.yml:150]: fixed

File: .github/workflows/release.yml (hunk 3 above)
Comment (shiki-tak) [reply .github/workflows/release.yml:86]: Thanks,  it is need.

File: .github/workflows/builds_and_tests.yml (hunk 2 above)
Comment (tnasu) [reply .github/workflows/builds_and_tests.yml:152]: How about leaving the job name only? (No need to write the comment.)
```
      - name: Build & Test static library
```

File: .github/workflows/release.yml (hunk 4 above)
Comment (tnasu) [reply .github/workflows/release.yml:54]:
```
      - name: Build & Test static library
```

File: .github/workflows/builds_and_tests.yml (hunk 1 above)
Comment (tnasu) [reply .github/workflows/builds_and_tests.yml:138]: Then, could we run `build_shared_library` and `build_static_lib` in parallel? As in my checking, it seems that `build_static_lib` doesn't have the dependency on `build_shared_library`.

File: .github/workflows/release.yml (hunk 4 above)
Comment (tnasu) [reply .github/workflows/release.yml:54]: Would you update this?

**Review by tnasu — APPROVED** [review]

**Comment by tnasu** [comment]
This is failed.
* https://github.com/line/wasmvm/runs/7834243083?check_suite_focus=true

```
Step 5/16 : RUN wget https://musl.cc/aarch64-linux-musl-cross.tgz   && tar -xf aarch64-linux-musl-cross.tgz   && mv ./aarch64-linux-musl-cross /opt   && /opt/aarch64-linux-musl-cross/bin/aarch64-linux-musl-gcc --version   && rm aarch64-linux-musl-cross.tgz
 ---> Running in 0d2cc39a2f5e
Connecting to musl.cc (104.232.42.245:443)
wget: can't connect to remote host (104.232.42.245): Operation timed out
The command '/bin/sh -c wget https://musl.cc/aarch64-linux-musl-cross.tgz   && tar -xf aarch64-linux-musl-cross.tgz   && mv ./aarch64-linux-musl-cross /opt   && /opt/aarch64-linux-musl-cross/bin/aarch64-linux-musl-gcc --version   && rm aarch64-linux-musl-cross.tgz' returned a non-zero code: 1
make: *** [Makefile:13: docker-image-alpine] Error 1
```

Does this relate the below? 
* https://github.com/orgs/community/discussions/27906#discussioncomment-3332440

Should we prepare a proxy server?

**Comment by shiki-tak** [comment]
> 

I try to re make the Dockerfile using docker's image without using wget.

**Comment by shiki-tak** [comment]
I need aarch64-linux-musl-cross for the wasmer build, but the github actions refuse to get it from [musl.cc](https://musl.cc/) using wget.

Therefore, I use the image that already contains this as a builder and copy and use it.
https://github.com/line/wasmvm/pull/71/files#diff-a3e40ace4c88526a8320154b12cfc19c7b048c6fd2ea67ddddb3e4849711f249R10

If we should prepare our own Docker image beforehand without using this, then I think this is a matter for another issue.

File: .github/workflows/builds_and_tests.yml [hunk 6]
```diff
@@ -134,6 +134,21 @@ jobs:
           name: libwasmvm
           path: ./api/libwasmvm.x86_64.so
 
+  build_static_lib:
+    runs-on: ubuntu-latest
+    steps:
+      - name: set up
+        uses: actions/setup-go@v2
+        with:
+          go-version: ^1.17.4
```
Comment (zemyblue) [inline .github/workflows/builds_and_tests.yml:143]: We need to upgrade to 1.18 via another PR

File: .github/workflows/builds_and_tests.yml [hunk 7]
```diff
@@ -134,6 +134,21 @@ jobs:
           name: libwasmvm
           path: ./api/libwasmvm.x86_64.so
 
+  build_static_lib:
+    runs-on: ubuntu-latest
+    steps:
+      - name: set up
+        uses: actions/setup-go@v2
```
Comment (zemyblue) [inline .github/workflows/builds_and_tests.yml:141]: The latest version of `setup-go` is 3.2.1. Please check it.

File: builders/Dockerfile.alpine [hunk 8]
```diff
@@ -1,3 +1,14 @@
+# Install C compiler for cross-compilation. This is required by
+# Wasmer in https://github.com/wasmerio/wasmer/blob/2.2.1/lib/vm/build.rs.
+# For newer versions this might not be needed anymore since build.rs is removed
+# in https://github.com/wasmerio/wasmer/pull/2807.
+#
+# https://unix.stackexchange.com/questions/620205/aarch64-linux-musl-cross-has-a-broken-link-for-ld-musl-aarch64-so-1
+
+# We need aarch64-linux-musl-cross to compile, wget to musl.cc is denied by github actions.
+# We will use a copy from an image that already contains aarch64-linux-musl-cross.
+FROM mondoolabs/golang:1.15.0-alpine AS builder
```
Comment (zemyblue) [inline builders/Dockerfile.alpine:10]: Please check the golang version.

Comment (shiki-tak) [reply builders/Dockerfile.alpine:10]: This image is only used to copy aarch64-linux-musl-cross. Therefore, the golang version of this image does not affect the build result.

File: .github/workflows/builds_and_tests.yml (hunk 6 above)
Comment (shiki-tak) [reply .github/workflows/builds_and_tests.yml]: fixed

File: .github/workflows/builds_and_tests.yml (hunk 7 above)
Comment (shiki-tak) [reply .github/workflows/builds_and_tests.yml]: fixed

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -34,6 +34,39 @@ jobs:
...
+
+  build_static_lib:
+    name: Build static libraries
+    needs:
+      - get-version
+    if: ${{ needs.get-version.outputs.package-version != needs.get-version.outputs.latest-tag }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: set up
+        uses: actions/setup-go@v2
```
Comment (zemyblue) [inline .github/workflows/release.yml:46]: Also use v2. The latest version is 3.2.1

Comment (shiki-tak) [reply .github/workflows/release.yml]: fixed

**Review by zemyblue — APPROVED** [review]

**Pull request merged**